### ISSUE TITLE
NullPointerException while using findCount with forUpdate

### ISSUE THREAD

**Issue by AntoineDuComptoirDesPharmacies** [ebean-orm/ebean#2762]
When we use Ebean, we sometimes use the same query to get paginated records from database and count the total number of rows matching the predicate.
In specific situation, this query can have a "forUpdate" flag to lock the paginated records returned.

## Expected behavior

In this situation, we expect the records query to lock the rows and the count query to return the correct number of element without considering the forUpdate lock.

## Actual behavior

NullPointerException occurs while trying to fetch the count.

### Steps to reproduce

```java
        DB.find(DelayedJobQueue.class)
                .forUpdate()
                .setFirstRow(0)
                .setMaxRows(10)
                .findPagedList().getTotalCount();
```

```console
javax.persistence.PersistenceException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at io.ebeaninternal.server.query.LimitOffsetPagedList.getTotalCount(LimitOffsetPagedList.java:101)
        at utils.db.searchResult.executors.BaseQueryExecutor.lambda$doExecute$2(BaseQueryExecutor.java:91)
        at utils.db.searchResult.executors.BaseQueryExecutor.doExecute(BaseQueryExecutor.java:99)
        at utils.db.searchResult.executors.QueryExecutor.execute(QueryExecutor.java:10)
        at domain.delayedJobQueue.book.DelayedJobQueueDatabaseBook.find(DelayedJobQueueDatabaseBook.java:84)
        at domain.delayedJobQueue.book.DelayedJobQueueDatabaseBook.getNextJobInQueue(DelayedJobQueueDatabaseBook.java:33)
        at service.JobService.buildInvoicesZip(JobService.java:233)
        at provide.job.controllers.JobApiController.buildInvoicesZip(JobApiController.java:43)
        at job.Routes$$anonfun$routes$1.$anonfun$applyOrElse$75(Routes.scala:1102)
        at play.core.routing.HandlerInvokerFactory$$anon$8.resultCall(HandlerInvoker.scala:150)
        at play.core.routing.HandlerInvokerFactory$$anon$8.resultCall(HandlerInvoker.scala:149)
        at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115) 
        at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
        at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:33)
        at be.objectify.deadbolt.java.actions.AbstractDeadboltAction.authorizeAndExecute(AbstractDeadboltAction.java:280)
        at be.objectify.deadbolt.java.ConstraintLogic.pass(ConstraintLogic.java:475)
        at be.objectify.deadbolt.java.ConstraintLogic.lambda$restrict$5(ConstraintLogic.java:141)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:143)
        at be.objectify.deadbolt.java.ConstraintLogic.restrict(ConstraintLogic.java:129)
        at be.objectify.deadbolt.java.actions.RestrictAction.applyRestriction(RestrictAction.java:73)
        at be.objectify.deadbolt.java.actions.AbstractRestrictiveAction.lambda$null$1(AbstractRestrictiveAction.java:59)
        at java.base/java.util.Optional.orElseGet(Optional.java:369)
        at be.objectify.deadbolt.java.actions.AbstractRestrictiveAction.lambda$execute$2(AbstractRestrictiveAction.java:59)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:143)
        at be.objectify.deadbolt.java.actions.AbstractRestrictiveAction.execute(AbstractRestrictiveAction.java:58)
        at be.objectify.deadbolt.java.actions.AbstractDeadboltAction.call(AbstractDeadboltAction.java:121)
        at actions.SentryOn5XXAction.call(SentryOn5XXAction.java:39)
        at actions.NoCookiesAction.call(NoCookiesAction.java:26)
        at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175)
        at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
        at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
        at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
        at play.core.j.HttpExecutionContext.execute(HttpExecutionContext.scala:59)
        at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
        at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
        at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
        at scala.concurrent.impl.Promise$DefaultPromise.map(Promise.scala:182)
        at scala.concurrent.Future$.apply(Future.scala:678)
        at play.core.j.JavaAction.apply(JavaAction.scala:176)
        at play.api.mvc.Action.$anonfun$apply$4(Action.scala:82)
        at play.api.libs.streams.StrictAccumulator.$anonfun$mapFuture$4(Accumulator.scala:168)
        at scala.util.Try$.apply(Try.scala:210)
        at play.api.libs.streams.StrictAccumulator.$anonfun$mapFuture$3(Accumulator.scala:168)
        at scala.Function1.$anonfun$andThen$1(Function1.scala:85)
        at scala.Function1.$anonfun$andThen$1(Function1.scala:85)
        at play.api.libs.streams.StrictAccumulator.run(Accumulator.scala:199)
        at play.core.server.AkkaHttpServer.$anonfun$runAction$4(AkkaHttpServer.scala:423)
        at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
        at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$3(FastFuture.scala:51)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at io.ebeaninternal.server.query.BaseFuture.get(BaseFuture.java:30)
        at io.ebeaninternal.server.query.LimitOffsetPagedList.getTotalCount(LimitOffsetPagedList.java:99)
        ... 63 common frames omitted
Caused by: java.lang.NullPointerException: null
        at io.ebean.platform.postgres.PostgresPlatform.lock(PostgresPlatform.java:141)
        at io.ebean.platform.postgres.PostgresPlatform.withForUpdate(PostgresPlatform.java:136)
        at io.ebean.config.dbplatform.DatabasePlatform.completeSql(DatabasePlatform.java:717)
        at io.ebeaninternal.server.query.CQueryBuilder$BuildReq.buildSql(CQueryBuilder.java:732)
        at io.ebeaninternal.server.query.CQueryBuilder.buildSql(CQueryBuilder.java:519)
        at io.ebeaninternal.server.query.CQueryBuilder.buildRowCountQuery(CQueryBuilder.java:276)
        at io.ebeaninternal.server.query.CQueryEngine.findCount(CQueryEngine.java:162)
        at io.ebeaninternal.server.query.DefaultOrmQueryEngine.findCount(DefaultOrmQueryEngine.java:90)
        at io.ebeaninternal.server.core.OrmQueryRequest.findCount(OrmQueryRequest.java:350)
        at io.ebeaninternal.server.core.DefaultServer.findCountWithCopy(DefaultServer.java:1236)
        at io.ebeaninternal.server.query.CallableQueryCount.call(CallableQueryCount.java:28)
        at io.ebeaninternal.server.query.CallableQueryCount.call(CallableQueryCount.java:12)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.ebeaninternal.server.executor.DefaultBackgroundExecutor.lambda$logExceptions$0(DefaultBackgroundExecutor.java:77)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        ... 3 common frames omitted
```

Yours faithfully,
LCDP

**Comment by rbygrave** [comment]
> this query can have a "forUpdate" flag to lock the paginated records returned.

All `forUpdate` queries should occur in explicit transaction scope and the rows are only locked for that given transaction scope. In this sense it is somewhat odd to use paginated queries with _forUpdate_ (as paginated queries are typically only used for results that would go to the user interface.  

That is, the motivation for total count (and hence pagination query) is to to supply that to the user interface.  Yet forUpdate lock should be only hence for the transaction which should be a very short time.

Are you confident that you are using **forUpdate** correctly here?  Case you explain the use case?

-----
Yes, we can treat this as a bug and fix it but my concern is that forUpdate is being incorrectly used here (perhaps it is not doing what you think it's doing?)  We use optimistic locking for detecting changes over "user think time" and we do not use database locks for that etc.

**Comment by AntoineDuComptoirDesPharmacies** [comment]
Thank you for this answer @rbygrave.

I totally agree with you about the semantic distance existing between `forUpdate` and getting total count of rows from query.
In fact, we arrived to this result because we built a `helper` that automatically take a domain search POJO and execute the getRecords + getTotalCount + other features, all-in-one.
This helper is particularly good for our API served to users, returning the total count of records for pagination purpose.

Our bad practice may have be to use it for internal purpose to be faster in development but in fact we do not require the total count for internal usage. We "workaround" this NPE by bypassing this `helper` and executing the query through simple manner.
In fact, it is a better writing for internal use case with less useless database charge.

Whatever, I am wondering if it could be better to have either a custom exception saying that we can not do this action or ignore the `forUpdate` in count query. It will help future developers to find the root cause instead of just getting NPE.

Yours faithfully,
LCDP
